modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/block/BlockService.scala

```diff
@@ -81,13 +80,25 @@ object BlockService {
 
         addressBalanceChanges.toList.traverse {
           case (address, balanceChange) =>
-            for {
-              current <- addressStorage.getBalance(address)
-              newBalance <- refineV[NonNegative]
-                .apply(current.coerce + balanceChange)
-                .bimap(BalanceNegative, Balance(_))
+            val positive = balanceChange > 0
+            def getDiff =
+              NonNegLong
+                .from(if (positive) balanceChange else -balanceChange)
+                .leftMap(new Throwable(_))
                 .liftTo[F]
-            } yield (address, newBalance)
+
+            addressStorage
+              .getBalance(address)
+              .flatMap { currentBalance =>
+                getDiff.map(Balance(_)).map { diff =>
+                  val updated =
+                    if (positive) currentBalance.plus(diff) else currentBalance.minus(diff)
+
+                  updated.getOrElse(Balance.empty)
```

Comment:
Are you defaulting to 0 if balance gets negative? For me it should fail fast and throw an error, we can't have such a situation at this point. If we do we have a flaw in our code that needs to come up.

Reply:
fixed